### DISCUSSION TITLE
How to read+write to a character device without geting stuck ?

### DISCUSSION THREAD

**Discussion by zcattacz** [micropython/micropython#11830] [Unix & Windows]
In python `os` module can enable read+write to character device like below
```
try:
    d = os.open("/dev/usb/lp0", os.O_RDWR)
    os.write(d, "~!@")
    printer_status = os.read(d, 100) # returns less if remaining buffer is less than 100, without blocking
finally:
    os.close(d)
```
I wonder how should this be handled in micropython?

EDIT: @dlech kindly suggested `"w+"` flag, which opens the device correctly, now my new issue becomes if the response has no end marker and size is unknown how should the device be read() without getting blocked.

```
with open("/dev/usb/lp0", "w+b") as f:
    buf1 = bytearray(sz)
    f.readinto(buf1) # this blocks forever, if sz > remaining buffer size

    buf2 = b''
    while True:
        c += f.read(1) # this blocks forever if read over end of response.
        buf2 += c
        if c == end_marker: # in this case, which is unknown 
            break

   # stuck free scheme should be ?
```

**Comment by dlech** [comment]
You can just use `with open("/dev/usb/lp0", "w+") as f:` as you also can in CPython.

**Comment by zcattacz** [comment]
Hi @dlech, 
Thank you for the advice. The `"w+"` (or `"w+b"`) mode opens the device without error, but .read() call blocks forever, not returning anything. This behavior is reproducible in CPython3. So far it seems only `os.open()/os.write()/os.write()` works and successfully returns code from printer without delay.

**Reply by dlech** [reply]
If you know how many bytes you are expecting, you could try passing that as a parameter to `read()`.

**Reply by zcattacz** [reply]
Hi @dlech,
My test code in CPython and micropython is as below.
```
with open("/dev/usb/lp0", "w+b") as f:
    f.write(b"\r\n")
    f.write(b"~!@")
    print("written, ok")
    print(f.read(100).hex()) # here blocks forever
```

**Reply by zcattacz** [reply]
OK, I tried .read(), which blocks, I tried read(100), which also blocks, but works in os.read() in CPython.
os.read(dev, bytes) allows setting larger byte size and return less.

Hmm, with read(4) it works. so my question becomes:
How should I avoid getting stuck when calling filehandle.read() if the response is of variable length and has no end of response marker?

**Reply by dlech** [reply]
You could use `poll()` to know if there is data available to read one byte and poll again.

Or you could set the file to non-blocking mode. 

https://github.com/micropython/micropython-lib/tree/master/unix-ffi contains `fcntl` for setting the mode to non-blocking and also has `os.read()`.

**Reply by zcattacz** [reply]
Hi, could you elaborate on the use of `poll()`, do you mean `select.poll()` ?

I checked `micropython-pyserial` and `micropython-serial` repo, both use `fcntl`. 
I wonder if there is any out-of-the-box way, since I am not doing low-level setup like baud-rate, flow control etc.

**Reply by dlech** [reply]
> do you mean select.poll() ?

yes


> could you elaborate on the use of `poll()`

https://docs.micropython.org/en/latest/library/select.html

If the timeout is set to 0, then it does not block.

> I wonder if there is any out-of-the-box way,

Although probably not very efficient, polling and reading one byte at a time should work "out of the box" without fcntl.
